src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/ReflectionJsonFormatWriter.cs

```diff
@@ -0,0 +1,250 @@
+                        _reflectionClassWriter.ReflectionWriteStartElement(jsonWriter, key.ToString());
+                        _reflectionClassWriter.ReflectionWriteValue(jsonWriter, context, value.GetType(), value, false, primitiveContractForParamType: null);
+                        _reflectionClassWriter.ReflectionWriteEndElement(jsonWriter);
+                    }
+                }
+                else
+                {
+                    ReflectionWriteArrayAttribute(jsonWriter);
+
+                    PrimitiveDataContract primitiveContractForType = PrimitiveDataContract.GetPrimitiveDataContract(enumeratorReturnType);
+                    if (primitiveContractForType != null && primitiveContractForType.UnderlyingType != Globals.TypeOfObject)
+                    {
+                        while (enumerator.MoveNext())
+                        {
+                            object current = enumerator.Current;
+                            context.IncrementItemCount(1);
+                            primitiveContractForType.WriteXmlElement(jsonWriter, current, context, itemName, null /*namespace*/);
+                        }
+                    }
+                    else
+                    {
+                        bool isDictionary = collectionContract.Kind == CollectionKind.Dictionary || collectionContract.Kind == CollectionKind.GenericDictionary;
+                        while (enumerator.MoveNext())
+                        {
+                            object current = enumerator.Current;
+                            context.IncrementItemCount(1);
+                            _reflectionClassWriter.ReflectionWriteStartElement(jsonWriter, itemName);
+                            if (isDictionary)
+                            {
+                                var itemContract = XmlObjectSerializerWriteContextComplexJson.GetRevisedItemContract(collectionContract.ItemContract);
+                                var jsonDataContract = JsonDataContract.GetJsonDataContract(itemContract);
+
+                                jsonDataContract.WriteJsonValue(jsonWriter, current, context, collectionContract.ItemType.TypeHandle);
+                            }
+                            else
+                            {
+                                _reflectionClassWriter.ReflectionWriteValue(jsonWriter, context, enumeratorReturnType, current, false, primitiveContractForParamType: null);
+                            }
+
+                            _reflectionClassWriter.ReflectionWriteEndElement(jsonWriter);
+                        }
+                    }
+                }
+            }
+        }
+
+        private void ReflectionWriteObjectAttribute(XmlWriterDelegator xmlWriter)
+        {
+            xmlWriter.WriteAttributeString(
+                null /* prefix */,
+                JsonGlobals.typeString /* local name */,
+                null /* namespace */,
+                JsonGlobals.objectString /* value */);
```

Comment:
Nit: Named parameters instead of the inline comments? e.g.:

```c#
xmlWriter.WriteAttributeString(
    prefix: null,
    localName: JsonGlobals.typeString,
    ns: null,
    value: JsonGlobals.objectString);
```